### ISSUE TITLE
Dockerfile syntax highlighting for escaped quotes is not quite right

### ISSUE THREAD

**Issue by alexr00** [moby/moby#40114]
From @bwateratmsft in https://github.com/microsoft/vscode/issues/82988

Copy from @dlech original issue here: https://github.com/microsoft/vscode-docker/issues/75

> Given the following dockerfile snippet...
>
```
# setup a new user
COPY compiler.sudoers /etc/sudoers.d/compiler
RUN chmod 0440 /etc/sudoers.d/compiler && \
    adduser --disabled-password --gecos \"\" compiler && \
    usermod -a -G sudo compiler
USER compiler
WORKDIR /home/compiler
CMD ["/bin/bash"] ["--login"]
```

> The escaped quotes after `gecos` appear to throw off the syntax highlighting.
>
> ![image](https://user-images.githubusercontent.com/36966225/67213825-b79f1b80-f3ec-11e9-9caa-6cc4e1326ab3.png)

**Comment by riverar** [comment]
Any updates on this?

**Comment by rcjsuen** [comment]
@riverar Do you have a Dockerfile to share that exhibits the problem? What editor are you using?

**Comment by riverar** [comment]
Repro steps:
1. Create a Dockerfile in Visual Studio Code (latest)
2. Explicitly set the escape character to backtick `
3. Add a RUN statement that uses backslashes
4. Observe syntax highlighting does not appear as expected

Screenshot:
![image](https://user-images.githubusercontent.com/475132/77575007-1848ed80-6e90-11ea-8b94-6635620ffbd5.png)

**Comment by rcjsuen** [comment]
@riverar Thank you for sharing your files. I am working on semantic highlighting for Dockerfiles so I am hopeful that this will not be a problem for Visual Studio Code in the future after its semantic highlighting API has stabilized and it becomes a part of the language server protocol.

<img width="932" alt="image" src="https://user-images.githubusercontent.com/15629116/77577407-1fcabf80-6ead-11ea-8cb2-81c29c008398.png">

<img width="1102" alt="image" src="https://user-images.githubusercontent.com/15629116/77577537-5b658980-6ead-11ea-9e53-b60aae9badba.png">

**Comment by riverar** [comment]
Looks fantastic @rcjsuen, thanks! Any rough ETA to share?

**Comment by rcjsuen** [comment]
> Looks fantastic @rcjsuen, thanks! Any rough ETA to share?

@riverar Unfortunately, no. I do not work on Visual Studio Code so I have no idea when the semantic highlighting API (microsoft/vscode#86415) will be finalized. It will definitely be months rather than weeks though given that their release cadence is generally once a month.

Are you using the [Docker for Visual Studio Code extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker)?

**Comment by riverar** [comment]
@rcjsuen Yep, using the extension. Will keep an eye on that issue, thanks.

**Comment by dcormier** [comment]
It looks like the API has been finalized.

**Comment by bwateratmsft** [comment]
@rcjsuen hopefully should release with 1.44 then. :)

**Comment by rcjsuen** [comment]
> It looks like the API has been finalized.

@dcormier Yes, here are a list of things that will need to be done.

1. Visual Studio Code has to be released with the new API (though I imagine that is happening this week or next?).
1. Some changes may need to be made in the [Microsoft/vscode-languageserver-node repository](https://github.com/microsoft/vscode-languageserver-node) for the API to line up. I'm not 100% sure about this.
1. I need to make any necessary updates to my code and then cut new releases accordingly.
1. I need to create a pull request to the [`vscode-docker` extension](https://github.com/microsoft/vscode-docker) with my new changes. They need to be okay with upping their version requirement to the latest VS Code release. They may or may not have requirements to support older versions of VS Code, I have no idea. @bwateratmsft Can you or someone else pitch in here and comment if possible?
1. After the pull request is merged, Microsoft has to push the button to release a new version of their Docker extension. @bwateratmsft Again, will need you or someone else to comment and share an expected release timeline here if possible.

In the meantime, if anyone has any broken Dockerfiles lying around, you can help me by testing it with my [online Dockerfile editor](https://rcjsuen.github.io/dockerfile-language-service/). Please open any issues you find in the [rcjsuen/dockerfile-language-service repository](https://github.com/rcjsuen/dockerfile-language-service), thank you.

**Comment by bwateratmsft** [comment]
> 4. I need to create a pull request to the vscode-docker extension with my new changes. They need to be okay with upping their version requirement to the latest VS Code release. They may or may not have requirements to support older versions of VS Code, I have no idea. @bwateratmsft Can you or someone else pitch in here and comment if possible?

We're planning to lock down and start testing for the 1.1 release ~April 13th. Uptake of latest VSCode versions tends to be pretty quick, but we should evaluate closer to then how many would be left behind if we enforce 1.44.

> 5. After the pull request is merged, Microsoft has to push the button to release a new version of their Docker extension. @bwateratmsft Again, will need you or someone else to comment and share an expected release timeline here if possible.

Tentatively (no guarantees!) April 21st-ish.

**Comment by rcjsuen** [comment]
> Any rough ETA to share?

For anyone using Visual Studio Code with the [Docker extension](https://marketplace.visualstudio.com/items?itemName=ms-azuretools.vscode-docker), you can subscribe to microsoft/vscode-docker#1840.

If you use an editor with LSP integration, you should see where they are in terms of implementing support for the (still-in-proposed-state) `textDocument/semanticTokens` request.

**Comment by bsousaa** [comment]
cc @crazy-max can we move these files to the BuildKit repo?
